### PR TITLE
`HTTPClient` now handles response deserialization + Introduced `NetworkError` and `BackendError`

### LINKED ISSUE #695: Refactor `HTTPClient`
The current implementation of `HTTPClient` is mostly a direct port of its objective-c counterpart, but its logic is somewhat contrived, and it's not actually taking advantage of the two dispatchqueues it's managing. 

This entire class could be refactored so that: 
- it uses the corresponding queue to send stuff concurrently or serially
- its logic is simplified by leveraging the queue to, well, queue tasks, as oppposed to managing _another_ queue of tasks manually

### PR DESCRIPTION
**Pull request by NachoSoto**
Closes #695 and finishes [CF-195]. 
Depends on ~~#1431, #1432, #1433, #1437~~.

## Goals:

- [x] Handle all requests / response / deserialization errors in `HTTPClient`. Clients of `HTTPClient` will only have to handle the "happy" path: #1431
- [x] `HTTPClient` shouldn't return `Result.success` with failed responses, forcing clients to verify the response is actually a failure: https://github.com/RevenueCat/purchases-ios/pull/1431
- [x] Abstract error response deserialization / error creation: #1427
- [x] Abstract attribute error parsing: #1427 
- [x] Move response deserialization to `HTTPClient` based on a `ResponseType: Decodable` type, so the completion block will simply return a `Result<HTTPResponse<ResponseType>, Error>`
- [x] `ETagManager` should store response `Data` instead of a deserialized `[String: Any]`
- [x] Improved error types in `HTTPClient` to fix tests: #1437
- [x] Dealt with non-backwards compatible `ETagManager: #1438

## Changes:
- Replaced `HTTPResponse`'s body from `[String: Any]` to a generic `HTTPResponseBody`.
- Created `HTTPResponseBody` to abstract `Decodable` and provide some default implementations for types like `Data,` `[String: Any]` (for backwards compatibility to types that aren't `Decodable` yet), and `Decodable` itself.
- New `HTTPResponse.Result` typealias (`Result<HTTPResponse<HTTPResponseBody>, Error>`) used everywhere. This will allow replacing `Error` with a more specific `Error` so we can forward known typed errors, and make sure that we don't end up with the wrong error type, or with a very complex error hierarchy and the details buried in `underlyingError`.
- Each layer (only a few for now) has its own error type: `NetworkError`, `BackendError`, `OfferingsManager.Error`.
- `HTTPClient` for example has to produce `NetworkError`, operations produce `BackendError`
- The parent `BackendError` can have specific errors like `.missingAppUserID`, but also be simply a child error `case networkError(NetworkError)`
- All of these conform to a `ErrorCodeConvertible`, so there is a single point of code that converts from simple and readable errors (like `BackendError.emptySubscriberAttributes`, `.unexpectedBackendResponse(.loginResponseDecoding)`) into errors with all the context using `ErrorUtils`
- Converted `DNSError` into `NetworkError.dnsError`. Its functionality remains unchanged.
- Removed `Backend.RCSuccessfullySyncedKey` and `ErrorDetails.finishableKey` in favor of tested properties on `NetworkError`

### Leftovers

There's a few things I've decided to not finish for now that we can turn into Jiras:
- [ ] `CustomerInfo` still does't conform to `Decodable` (manual deserialization is still supported by the current system): #1496
- `SubscriberAttributeDict` could also be `Codable`
- [x] Replace `OfferingsFactory` with `Decodable`: #1435

[CF-195]: https://revenuecats.atlassian.net/browse/CF-195?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Sources/Networking/HTTPClient.swift [hunk 1]
```diff
@@ -37,9 +37,9 @@ class HTTPClient {
         self.dnsChecker = dnsChecker
     }
 
-    func perform(_ request: HTTPRequest,
-                 authHeaders: [String: String],
-                 completionHandler: Completion?) {
+    func perform<Value: HTTPResponseBody>(_ request: HTTPRequest,
+                                          authHeaders: [String: String],
+                                          completionHandler: Completion<Value>?) {
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:42]: This is working now 🎉

File: Sources/Networking/HTTPClient.swift [hunk 2]
```diff
@@ -77,9 +77,26 @@ private extension HTTPClient {
 
         var httpRequest: HTTPRequest
         var headers: HTTPClient.RequestHeaders
-        var completionHandler: Completion?
+        var completionHandler: HTTPClient.Completion<Data>?
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:80]: This was the key. To be able to store a type erased `Request`, the completion handler just deals with the `Data` itself. The constructor is the only thing that has the `Value` type information, so it handles the deserialization there so everything else in `HTTPClient` is type agnostic.

File: Sources/Networking/HTTPClient.swift [hunk 3]
```diff
@@ -187,43 +201,51 @@ private extension HTTPClient {
...
+            if statusCode == .notModified {
+                return nil
             } else {
-                // No data if the body was empty, or if status is `.notModified`
-                // since the body will be fetched from the eTag.
-                return [:]
+                return data
             }
+
+//            if let data = data, statusCode != .notModified {
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:213]: This is one of the last TODOs, when the response fails to deserialize I need to improve how the error is created. Now the `CodableError` is what's forwarded which isn't ideal.

File: Sources/Networking/HTTPClient.swift [hunk 4]
```diff
@@ -187,43 +201,51 @@ private extension HTTPClient {
...
-                    Logger.error(Strings.network.json_data_received(dataString: dataAsString))
-
-                    throw jsonError
-                }
-
-                guard let result = result else {
-                    throw ErrorUtils.unexpectedBackendResponseError()
-                }
-
-                return result
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:209]: All this code is now gone 🎉

File: Sources/Networking/HTTPClient.swift [hunk 5]
```diff
@@ -355,3 +377,18 @@ private extension Encodable {
...
+
+    // Parses a `Result<HTTPResponse<Data>>` to `Result<HTTPResponse<Value>>`
+    func parseResponse<Value: HTTPResponseBody>() -> HTTPResponse<Value>.Result {
+        return self.flatMap { response in
+            HTTPResponse<Value>.Result {
+                try response.mapBody { data in
+                    try Value.create(with: data)
+                }
+            }
+        }
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:391]: I understand this code might be hard to follow, it's peak "functional" code. Ultimately it's just converting from `Data` to `Value` by calling the `HTTPResponseBody` decoding method. The nesting is because it's inside of the hierarchy of types of `Result` and `HTTPResponse`.

Feedback welcome if there's objections about it.

File: Sources/Networking/HTTPResponse.swift [hunk 6]
```diff
@@ -14,19 +14,44 @@
 
 import Foundation
 
-struct HTTPResponse {
+struct HTTPResponse<Body: HTTPResponseBody> {
 
-    typealias Body = [String: Any]
+    typealias Result = Swift.Result<Self, Error>
```
Comment (NachoSoto) [inline Sources/Networking/HTTPResponse.swift:19]: This `typealias` is used everywhere. This will allow replacing `Error` with a more specific `Error` so we can forward known typed errors, and make sure that we don't end up with the wrong error type, or with a very complex error hierarchy and the details buried in `underlyingError`s.

File: Sources/Networking/IntroEligibilityResponse.swift [hunk 7]
```diff
@@ -16,7 +16,7 @@ import Foundation
 // All parameters that are required to process the reponse from a GetIntroEligibility API call.
 struct IntroEligibilityResponse {
 
-    let result: Result<HTTPResponse, Error>
+    let result: Result<HTTPResponse<[String: Any]>, Error>
```
Comment (NachoSoto) [inline Sources/Networking/IntroEligibilityResponse.swift:19]: The new implementation still allows extracting a plain `[String: Any]` to maintain compatibility for types that aren't `Decodable` yet.

File: Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift [hunk 8]
```diff
@@ -17,33 +17,34 @@ class CustomerInfoResponseHandler {
...
-            }
-            .flatMap { response, info in
-                if !errorResponse.attributeErrors.isEmpty {
-                    return .failure(errorResponse.asBackendError(with: response.statusCode))
+            .flatMap {
+                if $0.body.errorResponse.attributeErrors.isEmpty {
+                    return .success($0.body.customerInfo)
                 } else {
-                    return .success(info)
+                    return .failure($0.body.errorResponse.asBackendError(with: $0.statusCode))
```
Comment (NachoSoto) [inline Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:27]: This is super simple now. If there's no attribute errors, return customer info. If there are attribute errors, return error.

File: Sources/Networking/Operations/LogInOperation.swift [hunk 9]
```diff
@@ -60,28 +61,11 @@ private extension LogInOperation {
...
-                    Logger.error(Strings.backendError.customer_info_instantiation_error(response: response))
-
-                    let extraContext = "statusCode: \(statusCode)"
-                    let subErrorCode = UnexpectedBackendResponseSubErrorCode
-                        .loginResponseDecoding
-                        .addingUnderlyingError(customerInfoError)
-                    let responseError = ErrorUtils.unexpectedBackendResponse(withSubError: subErrorCode,
-                                                                             extraContext: extraContext)
-                    return .failure(responseError)
-                }
```
Comment (NachoSoto) [inline Sources/Networking/Operations/LogInOperation.swift:84]: Goodbye custom deserialization code.

File: Tests/UnitTests/Networking/HTTPClientTests.swift [hunk 10]
```diff
@@ -281,33 +315,67 @@ class HTTPClientTests: XCTestCase {
...
+            result.value = response
         }
 
         expect(result.value).toEventuallyNot(beNil(), timeout: .seconds(1))
         expect(result.value).to(beFailure())
+
+        // TODO: custom error?
+        let error = try XCTUnwrap(result.value?.error as NSError?)
+        expect(error.domain) == RCPurchasesErrorCodeDomain
+        expect(error.code) == ErrorCode.apiEndpointBlockedError.rawValue
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/HTTPClientTests.swift:328]: This is a placeholder (this test fails). The last TODO to improve error handling will replace this with actual better errors.

File: Tests/UnitTests/Networking/HTTPClientTests.swift [hunk 11]
```diff
@@ -281,33 +315,67 @@ class HTTPClientTests: XCTestCase {
...
+                                     headers: nil)
+        }
+
+        self.client.perform(request, authHeaders: [:]) { (response: HTTPResponse<CustomResponse>.Result) in
+            result.value = response
+        }
+
+        expect(result.value).toEventuallyNot(beNil(), timeout: .seconds(1))
+        expect(result.value).to(beSuccess())
+        expect(result.value?.value?.body) == response
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/HTTPClientTests.swift:377]: The main test in the `HTTPClient` change, the body becomes the deserialized `CustomResponse` 🎉

**Comment by NachoSoto** [comment]
Marking this as ready for review because it's 95% ready. I might open a separate PR for the remaining 2 TODOs for easier review.

**Comment by NachoSoto** [comment]
Planning to finish #1437 to wrap this up this week.

File: Sources/Identity/CustomerInfo.swift [hunk 12]
```diff
@@ -19,7 +19,7 @@ import Foundation
  A container for the most recent customer info returned from `Purchases`.
  These objects are non-mutable and do not update automatically.
  */
-@objc(RCCustomerInfo) public class CustomerInfo: NSObject {
+@objc(RCCustomerInfo) public final class CustomerInfo: NSObject {
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:22]: > 🤔 is this necessary?

It is to be able to use `Self` in the `HTTPResponseBody` implementation. I could make that explicitly use `CustomerInfo`, but.

> I'm wondering if this might make it trickier for customers to use our classes for testing if they're doing inherit+replace

None of our classes are `open`. Since [Swift 3, Swift classes are sealed (non-open) by default](https://forums.swift.org/t/proposal-sealed-classes-by-default/3164). So this being `final` only affects the SDK itself.

File: Sources/Identity/CustomerInfo.swift [hunk 13]
```diff
@@ -421,3 +421,12 @@ private extension CustomerInfo {
     }
 
 }
+
+// Fixme: delete when `CustomerInfo` is `Decodable`
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:425]: There's not, I can create a Jira for it.
It's one of the last classes that still needs the conversion. That's why I've made the new system still work with manual serialization.

File: Sources/Networking/HTTPClient.swift (hunk 5 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:391]: > Honestly, I had a tough time understanding this. I think the mixture of functional methods, generic types, conversion of result types, Result.init with the block passed in directly, deep nesting, and implicit returns was a bit too much for my brain to handle.

Totally fair.

How about explaining what each line does in English (plus maybe adding all explicit types)?

```swift
// Parses a `Result<HTTPResponse<Data>>` to `Result<HTTPResponse<Value>>`
func parseResponse<Value: HTTPResponseBody>() -> HTTPResponse<Value>.Result {
    return self.flatMap { response in          // Convert the `Result` type
        HTTPResponse<Value>.Result {           // Create a new `HTTPResponse<Value>`
            try response.mapBody { data in     // Convert the body of `HTTPResponse<Data>` from `Data` -> `Value`
                try Value.create(with: data)   // Decode `Data` into `Value`
            }
        }
    }
}
```

File: Sources/Networking/ETagManager.swift [hunk 14]
```diff
@@ -130,6 +124,8 @@ private extension ETagManager {
         return request.url?.absoluteString
     }
 
+    // TODO: delete old data since this isn't backwards compatible
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/ETagManager.swift:127]: Instead of deleting the old data, I added a test in #1438 to ensure that `ETagManager` fails gracefully and treats it as a cache miss. How does that sound?

File: Sources/Networking/HTTPClient.swift (hunk 3 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:213]: Working on it in #1437, that's the last piece to finish.

File: Sources/Networking/HTTPClient.swift [hunk 15]
```diff
@@ -233,7 +255,7 @@ private extension HTTPClient {
                 guard response.statusCode.isSuccessfulResponse else {
                     return .failure(
                         ErrorResponse
-                            .from(response.jsonObject)
+                            .from(response.body)
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:258]: I like that 👍🏻

File: Sources/Networking/HTTPClient.swift (hunk 5 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:391]: Something to comment on this, is that this code is doing just a data transformation, it's not like it has some weird logic that needs to be maintained. There really is only one way to implement this, maybe many ways of _writing it_, but they would all compile to the same thing.

File: Sources/Networking/HTTPClient.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:258]: Great suggestions, I updated this.

**Comment by NachoSoto** [comment]
The tests and lint are fixed in #1437, but I don't want to merge it here because it would make this even longer to review.

**Comment by NachoSoto** [comment]
Any pending comments here? Let me know if you prefer me to merge #1437 in here to take a final look at all together, or you're ok with me merging them separately.

**Comment by aboedo** [comment]
@NachoSoto perhaps we can wait until #1437 so we can get tests passing first?

**Comment by NachoSoto** [comment]
Sure, you want me to merge that in here once it's approved? It depends on this branch.

**Comment by aboedo** [comment]
sure! #1437 is approved so we should be ready to go

File: Sources/Identity/CustomerInfo.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Identity/CustomerInfo.swift:425]: Working on that in #1496.

**Comment by NachoSoto** [comment]
Done! This PR is now completely functional.

**Review by aboedo — APPROVED** [review]
